drm-howto/modeset-atomic.c

```diff
@@ -0,0 +1,1082 @@
+			 */
+			if (crtc >= 0) {
+				fprintf(stdout, "crtc %u found for encoder %u, will need full modeset\n",
+					crtc, conn->encoders[i]);;
+				drmModeFreeEncoder(enc);
+				dev->crtc_id = crtc;
+				dev->crtc_index = j;
+				return 0;
+			}
+		}
+
+		drmModeFreeEncoder(enc);
+	}
+
+	fprintf(stderr, "cannot find suitable crtc for connector %u\n",
+		conn->connector_id);
+	return -ENOENT;
+}
+
+static uint64_t get_property_value(int fd, drmModeObjectPropertiesPtr props, const char *name)
+{
+	drmModePropertyPtr prop;
+	uint64_t value;
+	int j;
+	
+	for (j = 0; j < props->count_props; j++) {
+		prop = drmModeGetProperty(fd, props->props[j]);
+		if (!strcmp(prop->name, name))
+			value = props->prop_values[j];
+		drmModeFreeProperty(prop);
+	}
+	return value;
+}
+
+/*
+ * TODO: need doc
+ */
+static int modeset_find_plane(int fd, struct modeset_dev *dev)
+{
+	drmModePlaneResPtr plane_res;
+	bool found_primary = false;
+	int i, ret = -EINVAL;
+
+	plane_res = drmModeGetPlaneResources(fd);
+	if (!plane_res) {
+		fprintf(stderr, "drmModeGetPlaneResources failed: %s\n", strerror(errno));
+		return -ENOENT;
+	}
+
+	for (i = 0; (i < plane_res->count_planes) && !found_primary; i++) {
+		int plane_id = plane_res->planes[i];
+
+		drmModePlanePtr plane = drmModeGetPlane(fd, plane_id);
+		if (!plane) {
+			fprintf(stderr, "drmModeGetPlane(%u) failed: %s\n", plane_id, strerror(errno));
+			continue;
+		}
+
+		if (plane->possible_crtcs & (1 << dev->crtc_index)) {
+			drmModeObjectPropertiesPtr props =
+				drmModeObjectGetProperties(fd, plane_id, DRM_MODE_OBJECT_PLANE);
+
+			/* primary or not, this plane is good enough to use. */
+			dev->plane_id = plane_id;
+			ret = 0;
+
+			/* Get the "type" property and check if this is a primary plane */
+			if (get_property_value(fd, props, "type") == DRM_PLANE_TYPE_PRIMARY)
+				/* found our primary plane, lets use that! */
+				found_primary = true;
+			
+			drmModeFreeObjectProperties(props);
+		}
+
+		drmModeFreePlane(plane);
+	}
+
+	drmModeFreePlaneResources(plane_res);
+
+	if (found_primary)
+		fprintf(stdout, "found primary plane, id: %d\n", dev->plane_id);
+	else
+		fprintf(stdout, "found non-primary plane, id: %d\n", dev->plane_id);
+	return ret;
+}
+/*
+ * modeset_create_fb() stays the same.
+ */
+
+static int modeset_create_fb(int fd, struct modeset_buf *buf)
+{
+	struct drm_mode_create_dumb creq;
+	struct drm_mode_destroy_dumb dreq;
+	struct drm_mode_map_dumb mreq;
+	int ret;
+
+	/* create dumb buffer */
+	memset(&creq, 0, sizeof(creq));
+	creq.width = buf->width;
+	creq.height = buf->height;
+	creq.bpp = 32;
+	ret = drmIoctl(fd, DRM_IOCTL_MODE_CREATE_DUMB, &creq);
+	if (ret < 0) {
+		fprintf(stderr, "cannot create dumb buffer (%d): %m\n",
+			errno);
+		return -errno;
+	}
+	buf->stride = creq.pitch;
+	buf->size = creq.size;
+	buf->handle = creq.handle;
+
+	/* create framebuffer object for the dumb-buffer */
+	ret = drmModeAddFB(fd, buf->width, buf->height, 24, 32, buf->stride,
```

Comment:
Might want to use AddFB2 for make the pixel format explicit, or add a comment about what pixel format exactly 24,32 means.

Reply:
Yes, please do always use `drmModeAddFB2.`

Reply:
Done.